src/input/isf.c

```diff
@@ -0,0 +1,796 @@
+static int parse_isf_header(GString *buf, struct context *inc)
+{
+	char *pattern, *data_section;
+	int ret, i;
+	size_t item_offset, data_section_offset;
+
+	if (inc == NULL)
+		return SR_ERR_ARG;
+
+	data_section = find_data_section(buf);
+	if (data_section == NULL)
+		return SR_ERR_DATA;
+	data_section_offset = (size_t) (data_section - buf->str);
+
+	/* Search for all header items. */
+	for (i = 0; i < HEADER_ITEMS_PARAMETERS; ++i) {
+		pattern = find_item(buf->str, data_section_offset, header_items[i]);
+		if (pattern == NULL) {
+			/* WFID is not required. */
+			if (i == WFID)
+				continue;
+			return SR_ERR_DATA;
+		}
+
+		/*
+		 * Calculate the offset of the header item value in the buffer
+		 * as well as its distance from the data section.
+		 */
+		item_offset = (size_t) (pattern - buf->str);
+		item_offset += strlen(header_items[i]);
+		if (item_offset >= data_section_offset)
+			return SR_ERR_DATA;
+
+		ret = process_header_item(buf->str + item_offset, data_section_offset - item_offset, inc, i);
+		if (ret != SR_OK)
+			return ret;
+	}
+
+	return SR_OK;
+}
+
+/*
+ * Check if the format matches ISF format.
+ *
+ * TODO: The header could be searched for more items
+ * aside from NR_PT to increase the confidence.
+ */
+static int format_match(GHashTable *metadata, unsigned int *confidence)
+{
+	const char default_extension[] = ".isf";
+	const char nr_pt[] = "NR_PT";
+
+	GString *buf;
+	char *fn;
+	size_t fn_len;
+
+	/* If the extension is '.isf', it is likely the format matches. */
+	fn = g_hash_table_lookup(metadata, GINT_TO_POINTER(SR_INPUT_META_FILENAME));
+	if (fn != NULL && (fn_len = strlen(fn)) >= strlen(default_extension)) {
+		if (strcmp(fn + fn_len - strlen(default_extension), default_extension) == 0) {
+			*confidence = 10;
+			return SR_OK;
+		}
+	}
+
+	buf = g_hash_table_lookup(metadata, GINT_TO_POINTER(SR_INPUT_META_HEADER));
+	/* Check if the header contains NR_PT item. */
+	if (buf == NULL || g_strstr_len(buf->str, buf->len, nr_pt) == NULL)
+		return SR_ERR;
+
+	/* The header contains NR_PT item, the confidence is high. */
+	*confidence = 50;
+	return SR_OK;
+}
+
+/* Initialize the ISF module. */
+static int init(struct sr_input *in, GHashTable *options)
+{
+	struct context *inc;
+
+	(void) options;
+
+	in->sdi = g_malloc0(sizeof(struct sr_dev_inst));
+	in->priv = g_malloc0(sizeof(struct context));
+
+	inc = in->priv;
+	inc->create_channel = TRUE;
+
+	return SR_OK;
+}
+
+/*
+ * Read an integer value from the data buffer.
+ * The number of bytes per sample may vary and the sample is stored
+ * in a signed 64-bit integer. Therefore a negative integer extension
+ * might be needed.
+ */
+static float read_int_sample(struct sr_input *in, size_t offset)
+{
+	struct context *inc;
+	unsigned int bytnr;
+	int i;
+	int64_t value;
+	uint8_t data[MAX_INT_BYTNR];
+
+	inc = in->priv;
+	bytnr = inc->bytnr;
+
+	/* Value bytnr is checked in function "receive". */
+	memcpy(data, in->buf->str + offset, bytnr);
```

Comment:
bounds check on bytnr before memcpy ? not sure if that's what you mean in the comment on the previous line
[EDIT] ok, yes I see now, it is already checked. But I would probably just replace that comment with the bounds check, even if it's redundant - very inexpensive precaution and 1% easier to review

Reply:
Added bounds checking on bytnr.